custom-api/store_server.py

```diff
@@ -0,0 +1,296 @@
+        format = Format[format],
+        disposition = Disposition.EXTERNAL_LINKS,
+        wait_timeout = "0s",
+        warehouse_id = warehouse_id,
+        parameters = parameters,
+        row_limit = row_limit,
+        byte_limit = 500000000 # 500MB limit because lcp crashes.
+        )
+
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
+    }
+
+    return response
+
+# Poll the Statement Execution API for status on a previously generated execution request.
+def get_list_sales_status(request_id):
+    statement_response = w.statement_execution.get_statement(statement_id = request_id)
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
+    }
+
+    if statement_response.status.state == StatementState.SUCCEEDED:
+        response['chunk_count'] = statement_response.manifest.total_chunk_count
+
+    return response
+
+# Get the external link for the given chunk. The pre-signed URL is valid for 15 mins.
+def get_list_sales_chunk(request_id, chunk_index):
+    result_data = w.statement_execution.get_statement_result_chunk_n(
+        statement_id = request_id,
+        chunk_index = chunk_index
+    )
+
+    response = {
+        'request_id': request_id,
+        'link': rewrite_external_link(result_data.external_links[0].external_link)
+    }
+
+    return response
+
+#
+# Endpoint to add a sale to a store.
+#
+# This endpoint uses a parameterized query to insert data in to the Lakehouse via the Statement Execution API.
+#
+# Arguments:
+#   store_id - Required ID of the store supplied in the URL.
+#   date - Required date of the day the sale took place.
+#   price - Required price of each item sold.
+#   quantity - Required number of items sold.
+#   item-id - Required ID for the item sold.
+#   
+@app.route("/api/1.0/stores/<string:store_id>/sales", methods=["POST"])
+def insert_sale(store_id):
+    sold_date = request.form.get('date')
+    sales_price = request.form.get('price')
+    quantity = request.form.get('quantity')
+    item_id = request.form.get('item-id')
+    sale_id = random.getrandbits(32)
+    parameters = [
+        StatementParameterListItem(name='sold_date', value=sold_date, type="DATE"),
+        StatementParameterListItem(name='sale_id', value=sale_id, type="BIGINT"),
+        StatementParameterListItem(name='store_id', value=store_id, type="INT"),
+        StatementParameterListItem(name='item_id', value=item_id, type="INT"),
+        StatementParameterListItem(name='quantity', value=quantity, type="INT"),
+        StatementParameterListItem(name='price', value=sales_price, type="DECIMAL(7,2)")
+    ]
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.INSERT_SALE],
+        wait_timeout = "50s",
+        on_wait_timeout = TimeoutAction.CONTINUE,
```

Comment:
QQ: I don't think we poll if this doesn't complete within 50s right?

Reply:
Nope. It's fire and forget 😄 . Lazy on the FE. Should probably poll and see it to a terminal state.